### PR TITLE
CMR-5355: Changed the algorithm for size estimate for ascii output.

### PR DESCRIPTION
**Pull request by sxu123**

### CODE REVIEW

File: project.clj
Comment (chris-durbin) [inline project.clj:33]: Didn't realize 1.10 was out of beta.

File: src/cmr/sizing/formats.clj
Comment (chris-durbin) [inline src/cmr/sizing/formats.clj]: Since you use this a couple of times it might be worth pulling out into an average function.

File: src/cmr/sizing/formats.clj
Comment (chris-durbin) [inline src/cmr/sizing/formats.clj:60]: If there are no fill values in the variable metadata you will end up dividing by zero. Are FillValues a required field in UMM-Var?

If they are optional we might need to just pick an expected fill value of say 3 digits.

File: src/cmr/sizing/formats.clj
Comment (chris-durbin) [inline src/cmr/sizing/formats.clj:87]: Similar comment about dividing by zero if ValidRanges are an optional field in the metadata.

If they are optional I think we should default to using the fill value here when not provided. If both fill value and valid ranges are not provided I'd probably default to 3 again.

File: src/cmr/sizing/formats.clj
Comment (chris-durbin) [inline src/cmr/sizing/formats.clj:112]: Awesome documentation.

File: src/cmr/sizing/formats.clj
Comment (chris-durbin) [inline src/cmr/sizing/formats.clj]: Oh because there's one less comma and space. I wouldn't have thought of that. 👍 

Nit: I'd probably use `dec` rather than subtract 1 for readability.

File: src/cmr/sizing/formats.clj
Comment (chris-durbin) [inline src/cmr/sizing/formats.clj]: Based on the unit tests it looks like we should cast this to a long (estimates should be in bytes, no decimal numbers).

File: test/cmr/sizing/tests/unit/formats.clj
Comment (chris-durbin) [inline test/cmr/sizing/tests/unit/formats.clj:1]: Good use of unit tests.

You will also need to update the service-bridge integration tests with the new ASCII estimates.

**Review by chris-durbin — COMMENTED** [review]
PR looks great. I just have questions about optional vs. mandatory UMM variables before merging this in.

Also there was a jar vulnerability reported in the Travis CI tests.

File: src/cmr/sizing/formats.clj
Comment (chris-durbin) [reply src/cmr/sizing/formats.clj:60]: I just double-checked - both Fill Values and ValidRanges are optional. We're going to have to figure out what to do when they are not provided.

https://git.earthdata.nasa.gov/projects/EMFD/repos/unified-metadata-model/browse/variable/v1.3/umm-var-json-schema.json
```
      "ValidRanges": {
        "description": "Valid ranges of variable data values.",
        "type": "array",
        "items": {
          "$ref": "#/definitions/ValidRangeType"
        },
        "minItems": 0
      },
```

```
      "FillValues": {
        "description": "The fill value of the variable in the data file. It is generally a value which falls outside the valid range. For example, if the valid range is '0, 360', the fill value may be '-1'. The fill value type is data provider-defined. For example, 'Out of Valid Range'.",
        "type": "array",
        "items": {
          "$ref": "#/definitions/FillValueType"
        },
        "minItems": 0
      }
```

File: project.clj
Comment (oubiwann) [reply project.clj:33]: Yeah, just happened yesterday.

File: src/cmr/sizing/formats.clj
Comment (oubiwann) [reply src/cmr/sizing/formats.clj]: Yeah, in fact, the linter should have caught that ... it's one of the idioms recommended by the Clojure style guide.

**Comment by oubiwann** [comment]
Looking good so far; +1 to all of Chris' feedback. 

Ping us when the changes have been made, and I'll do another review.

File: src/cmr/sizing/formats.clj
Comment (daniel-zamora) [inline src/cmr/sizing/formats.clj]: I can't find any validation on Size for umm var dimensions that makes sure its greater than zero, or not zero.  It probably wouldn't affect the total estimate too much but it might be something we want to protect against.

File: src/cmr/sizing/formats.clj
Comment (sxu123) [reply src/cmr/sizing/formats.clj:60]: Done.

File: test/cmr/sizing/tests/unit/formats.clj
Comment (sxu123) [reply test/cmr/sizing/tests/unit/formats.clj]: This is the part I don't know what to do. So it's a different repo, so I will have to submit a separate PR? Also, since the tests are run against sit, I won't be able to verify until after it's merged?

File: src/cmr/sizing/formats.clj
Comment (sxu123) [reply src/cmr/sizing/formats.clj:87]: Done.

File: test/cmr/sizing/tests/unit/formats.clj
Comment (sxu123) [inline test/cmr/sizing/tests/unit/formats.clj:59]: I didn't choose the magic 3, but it's magic.

File: src/cmr/sizing/formats.clj
Comment (sxu123) [inline src/cmr/sizing/formats.clj]: This doesn't count the special case of {:Min 0 :Max 0}. but better than not counting the negative to positive range at all.

**Comment by oubiwann** [comment]
One more thing that needs to be changed: with the move to Clojure 1.10, Java 7 is no longer supported, so you'll need to remove those from the `.travis.yml` file.

Come to think of it, Travis now has a fixed OpenJDK 8 in operation, so we can actually change the `openjdk7` line to `openjdk8`.

Note that while Clojure 1.10 supports (and encourages) JDK 11, the Travis CI Debian images haven't been updated with the JDK / Clojure fixes necessary for this (packaging updates need to be applied due to packaging bugs in older releases in Debian). As such, we can't add Oracle or OpenJDK 11.

**Review by oubiwann — APPROVED** [review]

**Review by chris-durbin — APPROVED** [review]
Approved assuming Travis CI file is updated to no longer build against JDK 7.

**Comment by sxu123** [comment]
> One more thing that needs to be changed: with the move to Clojure 1.10, Java 7 is no longer supported, so you'll need to remove those from the `.travis.yml` file.
> 
> Come to think of it, Travis now has a fixed OpenJDK 8 in operation, so we can actually change the `openjdk7` line to `openjdk8`.
> 
> Note that while Clojure 1.10 supports (and encourages) JDK 11, the Travis CI Debian images haven't been updated with the JDK / Clojure fixes necessary for this (packaging updates need to be applied due to packaging bugs in older releases in Debian). As such, we can't add Oracle or OpenJDK 11.

Thank you!

**Comment by oubiwann** [comment]
Oh, wait -- one more change is needed: `0.2.3-SNAPSHOT` needs to be bumped to `0.2.4-SNAPSHOT` in order for Nexus to pick up the latest from Clojars.

Once you merge this to master, I can publish it to Clojars, and once that's done, you can push an update to cmr-service-bridge that uses `0.2.4-SNAPSHOT` ...

**Pull request merged**
